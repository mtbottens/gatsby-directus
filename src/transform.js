import crypto from 'crypto';
import _ from 'lodash';
import Colors from 'colors';
import markdown from './transformers/markdown';
import image from './transformers/image';

/**
 * Returns string with first letter uppercased
 * 
 * @param  {String} str 
 */
const upperCase = str => 
    str.charAt(0).toUpperCase() + str.slice(1);

/**
 * Converts a string into a digest for Gatsby to interpret
 * 
 * @param {String} str 
 */
const digest = str =>
    crypto
        .createHash(`md5`)
        .update(str)
        .digest(`hex`);

/** Prevent these fields from directus overriding gatsby internal nodes */        
const RESTRICTED_NODE_FIELDS = [
    `id`,
    `children`,
    `parent`,
    `fields`,
    `internal`,
];

/** Prefix for directus node types */
const NODE_ID_PREFIX = `Directus`;

/**
 * Converts any number of arguments into a gatsby node id or type
 * 
 * @param  {String[]} ...identifiers
 */
const createNodeId = (...identifiers) => {
    let nodeId = `${NODE_ID_PREFIX}`;

    for (let identifier of identifiers) {
        nodeId += `__${identifier}`;
    }

    return nodeId;
};

/** Array of transformer objects used to convert data to corrent types */
const transformers = [markdown, image];

const transformer = ({ url, program, currentTables, createNode }) => {
    let entitiesCreated = 0;
    let EntityTypesArray = [];

    for (let table of currentTables) {
        // For fields that require no additional processing
        const basicFields = {};
        // For fields that need processing, ie. Adding mime/types for transformer plugins
        const complexFields = {};
        // Gets each entity from the table
        const entries = table.entries.slice();
        delete table.entries;

        for (let entity of entries) {
            const entityId = createNodeId(upperCase(table.name), entity.id);
            const entityType = createNodeId(upperCase(table.name));

            // Create a node for this entity
            const directusEntity = {
                id: entityId,
                parent: `__source__`,
                children: [],
                internal: {
                    type: entityType
                }
            };

            // Remove conflicting fields
            const validKeys = _.filter(Object.keys(entity), (key) => {
                return RESTRICTED_NODE_FIELDS.indexOf(key) === -1;
            });

            // Loop through each valid key and transform them into basic or complex nodes
            for (let validKey of validKeys) {
                const columnData = _.find(table.columns.data, { id: validKey });
                const value = entity[validKey];

                const columnTransformer = _.find(transformers, function(transformer) {
                    return transformer.test(columnData);
                });

                if (columnTransformer) {
                    const node = columnTransformer.transform({
                        url,
                        program,
                        table,
                        entity,
                        directusEntity,
                        validKey,
                        value,
                        createNodeId,
                        digest,
                    });
                    createNode(node);
                    directusEntity.children = directusEntity.children.concat([node.id]);
                    complexFields[`${validKey}___NODE`] = node.id;
                } else {
                    basicFields[validKey] = value;
                }
            }

            directusEntity = { ...complexFields, ...basicFields, ...directusEntity};
            directusEntity.internal.contentDigest = digest(JSON.stringify(directusEntity));
            createNode(directusEntity);

            // Stats for console
            if (EntityTypesArray.indexOf(entityType) === -1) EntityTypesArray.push(entityType);
            entitiesCreated++;
        }
    }

    // Output generation stats to console
    console.log(`Directus Entity Types Processed: ${EntityTypesArray.length}`.blue);
    console.log(`Directus Entities Created: ${entitiesCreated}`.blue);
};

export { transformer };