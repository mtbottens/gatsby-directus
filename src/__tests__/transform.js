import {transformer} from '../transform'
import simpleData from './data/simple-data.json'
import markdownData from './data/markdown-data.json'
import toggleData from './data/toggle-data.json'
import manyToOneData from './data/many-to-one-data.json'
import manyToManyData from './data/many-to-many-data.json'

describe(`Processes Data From Directus CMS`, () => {
    it(`Properly Builds Simple Tables`, async () => {
        let transformedData = await transformer({
            currentTables: simpleData,
            allTables: simpleData
        });
        expect(transformedData).toMatchSnapshot();
    })

    it(`Properly Processes Markdown Complex Fields`, async () => {
        let transformedData = await transformer({
            currentTables: markdownData,
            allTables: markdownData
        })
        expect(transformedData).toMatchSnapshot();
    })

    it(`Properly Processes Toggle Complex Fields`, async () => {
        let transformedData = await transformer({
            currentTables: toggleData,
            allTables: toggleData
        })
        for (let entry of transformedData.data['test_table']) {
            if (entry.id === 'Directus__Test_table__1') {
                expect(entry.toggle).toBeTruthy();
            } else {
                expect(entry.toggle).toBeFalsy();
            }
        }
        expect(transformedData).toMatchSnapshot();
    })

    it(`Properly Processes Many to One or One to Many Relationships`, async () => {
        let transformedData = await transformer({
            currentTables: manyToOneData,
            allTables: manyToOneData
        });
        // manyToOneData also contains markdown for nested complex fields
        expect(transformedData).toMatchSnapshot();
    })

    it(`Propery Processes Many to Many Relationships`, async () => {
        let transformedData = await transformer({
            currentTables: manyToManyData,
            allTables: manyToManyData
        });
        expect(transformedData).toMatchSnapshot();
    })
})