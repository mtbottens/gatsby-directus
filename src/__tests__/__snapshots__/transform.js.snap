// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Processes Data From Directus CMS Properly Builds Simple Tables 1`] = `
Object {
  "data": Object {
    "test_table": Array [
      Object {
        "children": Array [],
        "id": "Directus__Test_table__1",
        "internal": Object {
          "contentDigest": "eaeab362f5df52fdcd00c6f60e370629",
          "type": "Directus__Test_table",
        },
        "parent": "__source__",
        "somedata": "somevalue",
      },
    ],
  },
  "dependentNodeQueue": Array [],
}
`;

exports[`Processes Data From Directus CMS Properly Processes Many to One or One to Many Relationships 1`] = `
Object {
  "data": Object {
    "test_table": Array [
      Object {
        "children": Array [
          "Directus__Test_table__test_table_type__1",
        ],
        "id": "Directus__Test_table__1",
        "internal": Object {
          "contentDigest": "0e69c7aa9b28ca52b615eaab7cdfe4d4",
          "type": "Directus__Test_table",
        },
        "parent": "__source__",
        "somedata": "somevalue",
        "type___NODE": "Directus__Test_table__test_table_type__1",
      },
    ],
    "test_table_type": Array [
      Object {
        "children": Array [
          "Directus__Test_table_type__1__type_description",
        ],
        "id": "Directus__Test_table_type__1",
        "internal": Object {
          "contentDigest": "134247ec0bdf8735ee3277fa715aa5ed",
          "type": "Directus__Test_table_type",
        },
        "parent": "__source__",
        "type": "sometype",
        "type_description___NODE": "Directus__Test_table_type__1__type_description",
      },
    ],
  },
  "dependentNodeQueue": Array [
    Object {
      "children": Array [],
      "id": "Directus__Test_table__test_table_type__1__type_description",
      "internal": Object {
        "content": "# this type is the best type",
        "contentDigest": "b38ca283f92e383995a73e343405b5d8",
        "mediaType": "text/markdown",
        "type": "Directus__test_table__test_table_type__type_description",
      },
      "parent": "Directus__Test_table__test_table_type__1",
    },
    Object {
      "children": Array [
        "Directus__Test_table__test_table_type__1__type_description",
      ],
      "id": "Directus__Test_table__test_table_type__1",
      "internal": Object {
        "contentDigest": "b047a3c8de2ea749191460ee71f68691",
        "type": "Directus__Test_table__test_table_type",
      },
      "parent": "__source__",
      "type": "sometype",
      "type_description___NODE": "Directus__Test_table__test_table_type__1__type_description",
    },
    Object {
      "children": Array [],
      "id": "Directus__Test_table_type__1__type_description",
      "internal": Object {
        "content": "# this type is the best type",
        "contentDigest": "b38ca283f92e383995a73e343405b5d8",
        "mediaType": "text/markdown",
        "type": "Directus__test_table_type__type_description",
      },
      "parent": "Directus__Test_table_type__1",
    },
  ],
}
`;

exports[`Processes Data From Directus CMS Properly Processes Markdown Complex Fields 1`] = `
Object {
  "data": Object {
    "test_table": Array [
      Object {
        "children": Array [
          "Directus__Test_table__1__markdown",
        ],
        "id": "Directus__Test_table__1",
        "internal": Object {
          "contentDigest": "2e744fc7f674a025aef1e2d48f536716",
          "type": "Directus__Test_table",
        },
        "markdown___NODE": "Directus__Test_table__1__markdown",
        "parent": "__source__",
        "somedata": "somevalue",
      },
    ],
  },
  "dependentNodeQueue": Array [
    Object {
      "children": Array [],
      "id": "Directus__Test_table__1__markdown",
      "internal": Object {
        "content": "# some header",
        "contentDigest": "d9f15833f1ef5b9a47799d0dfbbcfc78",
        "mediaType": "text/markdown",
        "type": "Directus__test_table__markdown",
      },
      "parent": "Directus__Test_table__1",
    },
  ],
}
`;

exports[`Processes Data From Directus CMS Properly Processes Toggle Complex Fields 1`] = `
Object {
  "data": Object {
    "test_table": Array [
      Object {
        "children": Array [],
        "id": "Directus__Test_table__1",
        "internal": Object {
          "contentDigest": "e5622000abf22d49740bc670e37bcf6f",
          "type": "Directus__Test_table",
        },
        "parent": "__source__",
        "somedata": "somevalue",
        "toggle": true,
      },
      Object {
        "children": Array [],
        "id": "Directus__Test_table__2",
        "internal": Object {
          "contentDigest": "2867069400c698db5d384fdc18041ab9",
          "type": "Directus__Test_table",
        },
        "parent": "__source__",
        "somedata": "somevalue",
        "toggle": false,
      },
    ],
  },
  "dependentNodeQueue": Array [],
}
`;

exports[`Processes Data From Directus CMS Propery Processes Many to Many Relationships 1`] = `
Object {
  "data": Object {
    "test_table": Array [
      Object {
        "children": Array [],
        "id": "Directus__Test_table__1",
        "internal": Object {
          "contentDigest": "37411f9aebf5e25167e1553a17a30f25",
          "type": "Directus__Test_table",
        },
        "parent": "__source__",
        "somedata": "somevalue",
        "type": Array [
          Object {
            "children": Array [
              "Directus__Test_table__test_table_types__1__type_description",
            ],
            "id": "Directus__Test_table__test_table_types__1",
            "internal": Object {
              "contentDigest": "9fec628022927c8d82230a4577b413b4",
              "type": "Directus__Test_table__test_table_types",
            },
            "parent": "__source__",
            "type": "sometype",
            "type_description___NODE": "Directus__Test_table__test_table_types__1__type_description",
          },
          Object {
            "children": Array [
              "Directus__Test_table__test_table_types__2__type_description",
            ],
            "id": "Directus__Test_table__test_table_types__2",
            "internal": Object {
              "contentDigest": "278b36390d55789173ced85484aea9d8",
              "type": "Directus__Test_table__test_table_types",
            },
            "parent": "__source__",
            "type": "sometype",
            "type_description___NODE": "Directus__Test_table__test_table_types__2__type_description",
          },
        ],
      },
    ],
    "test_table_types": Array [
      Object {
        "children": Array [
          "Directus__Test_table_types__1__type_description",
        ],
        "id": "Directus__Test_table_types__1",
        "internal": Object {
          "contentDigest": "e6dd04ea6a8fcf7d53c12bd2a40d1bac",
          "type": "Directus__Test_table_types",
        },
        "parent": "__source__",
        "type": "sometype",
        "type_description___NODE": "Directus__Test_table_types__1__type_description",
      },
      Object {
        "children": Array [
          "Directus__Test_table_types__2__type_description",
        ],
        "id": "Directus__Test_table_types__2",
        "internal": Object {
          "contentDigest": "34f3cc16b182bf928523a39babe15469",
          "type": "Directus__Test_table_types",
        },
        "parent": "__source__",
        "type": "sometype",
        "type_description___NODE": "Directus__Test_table_types__2__type_description",
      },
    ],
  },
  "dependentNodeQueue": Array [
    Object {
      "children": Array [],
      "id": "Directus__Test_table__test_table_types__1__type_description",
      "internal": Object {
        "content": "# this type is the best type",
        "contentDigest": "b38ca283f92e383995a73e343405b5d8",
        "mediaType": "text/markdown",
        "type": "Directus__test_table__test_table_types__type_description",
      },
      "parent": "Directus__Test_table__test_table_types__1",
    },
    Object {
      "children": Array [],
      "id": "Directus__Test_table__test_table_types__2__type_description",
      "internal": Object {
        "content": "# no this is the best type for real",
        "contentDigest": "7bac2f6afdf3e937cc58a7794342f941",
        "mediaType": "text/markdown",
        "type": "Directus__test_table__test_table_types__type_description",
      },
      "parent": "Directus__Test_table__test_table_types__2",
    },
    Object {
      "children": Array [],
      "id": "Directus__Test_table_types__1__type_description",
      "internal": Object {
        "content": "# this type is the best type",
        "contentDigest": "b38ca283f92e383995a73e343405b5d8",
        "mediaType": "text/markdown",
        "type": "Directus__test_table_types__type_description",
      },
      "parent": "Directus__Test_table_types__1",
    },
    Object {
      "children": Array [],
      "id": "Directus__Test_table_types__2__type_description",
      "internal": Object {
        "content": "# no this is the best type for real",
        "contentDigest": "7bac2f6afdf3e937cc58a7794342f941",
        "mediaType": "text/markdown",
        "type": "Directus__test_table_types__type_description",
      },
      "parent": "Directus__Test_table_types__2",
    },
  ],
}
`;
